### ISSUE TITLE
question

### ISSUE THREAD

**Issue by geraldstanje** [sunhailin-Leo/triton-service-go#71]
hi,

i have a triton inference server (nvcr.io/nvidia/tritonserver:23.12-py3) running locally:
```
modelMetadataResponse: name:"ensemble_model" versions:"1" platform:"ensemble" inputs:{name:"TEXT" datatype:"BYTES" shape:-1} outputs:{name:"label" datatype:"BYTES" shape:-1} outputs:{name:"probabilities" datatype:"FP64" shape:-1 shape:2}
```

i want to write a golang client to connect to the triton inference server via http.

how can i implement the golang client using your library doing the following?
```
import numpy as np
import tritonclient.http

model_name = 'ensemble_model'
url = 'localhost:8000'
model_version = '1'

text_16 = 'test'
triton_client = tritonclient.http.InferenceServerClient(url=url, verbose=False)
assert triton_client.is_model_ready(model_name=model_name, model_version=model_version), f"model {model_name} not yet ready"

model_metadata = triton_client.get_model_metadata(model_name=model_name, model_version=model_version)
model_config = triton_client.get_model_config(model_name=model_name, model_version=model_version)

query = tritonclient.http.InferInput(name='TEXT', shape=(1,), datatype="BYTES")
label = tritonclient.http.InferRequestedOutput(name='label', binary_data=False)
probabilities = tritonclient.http.InferRequestedOutput(name='probabilities', binary_data=False)

query.set_data_from_numpy(np.asarray([text_16], dtype=object))
response = triton_client.infer(model_name=model_name, model_version=model_version, inputs=[query], outputs=[label,probabilities])
result_label = response.as_numpy('label')
result_probabilities = response.as_numpy('probabilities')
print("label:", result_label, ", probabilities:", result_probabilities)
```

**Comment by sunhailin-Leo** [comment]
@geraldstanje 
* You can follow https://github.com/sunhailin-Leo/triton-service-go?tab=readme-ov-file#usage
* Our code needs to implement three parts: input, output, and return result processing. Currently, we are only compatible with Bert's processing, if it is some other model, you can follow the logic of Bert's implementation to implement.
* Of course, we'd be pleased to have a developer provide some basic packaging to submit to our repository and refine it.

**Comment by geraldstanje** [comment]
@sunhailin-Leo thanks for quick feedback.

can you tell me what the following is used for? do i need to implement those handlers?
```
	InferCallback                   nvidia_inferenceserver.DecoderFunc
	GenerateModelInferRequest       GenerateModelInferRequest
	GenerateModelInferOutputRequest GenerateModelInferOutputRequest
```

**Comment by sunhailin-Leo** [comment]
> @sunhailin-Leo thanks for quick feedback.
> 
> can you tell me what the following is used for? do i need to implement those handlers?
> 
> ```
> 	InferCallback                   nvidia_inferenceserver.DecoderFunc
> 	GenerateModelInferRequest       GenerateModelInferRequest
> 	GenerateModelInferOutputRequest GenerateModelInferOutputRequest
> ```

* `GenerateModelInferRequest` and `GenerateModelInferOutputRequest ` correspond to input and output in `config.pbtxt`.
* `InferCallback`: This is the function you need to implement to handle the content returned by triton.

**Comment by geraldstanje** [comment]
@sunhailin-Leo need to handle the following and define those shapes?
```
modelMetadataResponse: name:"ensemble_model" versions:"1" platform:"ensemble" inputs:{name:"TEXT" datatype:"BYTES" shape:-1} outputs:{name:"label" datatype:"BYTES" shape:-1} outputs:{name:"probabilities" datatype:"FP64" shape:-1 shape:2}
```

**Comment by sunhailin-Leo** [comment]
> @sunhailin-Leo GenerateModelInferRequest needs to define the following?
> 
> ```
> inputs:{name:"TEXT" datatype:"BYTES" shape:-1} 
> ```
> 
> GenerateModelInferOutputRequest needs to define the following?
> 
> ```
> outputs:{name:"label" datatype:"BYTES" shape:-1} outputs:{name:"probabilities" datatype:"FP64" shape:-1 shape:2}
> ```

* https://github.com/sunhailin-Leo/triton-service-go/blob/main/test/bert_test.go#L27-L64

**Comment by geraldstanje** [comment]
hi,

get an error here:
```
package main

import (
    "fmt"
	"log"
    "time"
	"strings"

    "github.com/sunhailin-Leo/triton-service-go/v2/models"
    "github.com/sunhailin-Leo/triton-service-go/v2/nvidia_inferenceserver"
    "github.com/sunhailin-Leo/triton-service-go/v2/utils"

    "github.com/valyala/fasthttp"
    "google.golang.org/grpc"
    "google.golang.org/grpc/credentials/insecure"
)

const (
	DataSplitString     string = " ||| "
)

func testGenerateModelInferRequest() []*nvidia_inferenceserver.ModelInferRequest_InferInputTensor {
	return []*nvidia_inferenceserver.ModelInferRequest_InferInputTensor{
		{
			Name:     "TEXT",
			Datatype: "BYTES",
			//Shape:    []int64{1},
			//Data: []string{"foo"},
			//Content: &nvidia_inferenceserver.InferTensorContents{
			//	BytesContents: []byte{"foo"}, //[][]byte{[]byte("Your input text here")},
			//},
		},
	}
}

// testGenerateModelInferOutputRequest Triton Output
func testGenerateModelInferOutputRequest(params ...interface{}) []*nvidia_inferenceserver.ModelInferRequest_InferRequestedOutputTensor {
	return []*nvidia_inferenceserver.ModelInferRequest_InferRequestedOutputTensor{
		{
			Name: "label",
			//Parameters: map[string]*nvidia_inferenceserver.InferParameter{
				/*tBertModelRespBodyOutputBinaryDataKey: {
					ParameterChoice: &nvidia_inferenceserver.InferParameter_BoolParam{BoolParam: false},
				},
				tBertModelRespBodyOutputClassificationDataKey: {
					ParameterChoice: &nvidia_inferenceserver.InferParameter_Int64Param{Int64Param: 1},
				},
				*/
			//},
		},
		{
			Name: "probabilities",
		},
	}
}

// testModerInferCallback infer call back (process model infer data)
func testModerInferCallback(inferResponse interface{}, params ...interface{}) ([]interface{}, error) {
	fmt.Printf("inferResponse: %v\n", inferResponse)
	fmt.Printf("params: %v\n", params...)
	return nil, nil
}

type SetFitModelService struct {
	models.ModelService
}

func NewSetFitModelService(ms models.ModelService) SetFitModelService {
    return SetFitModelService{
        ModelService: ms,
    }
}

type OffsetsType struct {
	Start int
	End   int
}

// InputFeature Bert InputFeature.
type InputFeature struct {
	Text     []string   // origin text
	//Tokens   []string // token. like CLS/SEP after tokenizer
	//TokenIDs []int32  // input_ids
	//Mask     []int32  // input_mask
	//TypeIDs  []int32  // segment_ids
}

// InputObjects bert input objects for position record.
type InputObjects struct {
	Input    string
	//Tokens   []string
	//PosArray []OffsetsType
}

// HTTPBatchInput Model HTTP Batch Request Input Struct (Support batch 1).
type HTTPBatchInput struct {
	Name     string  `json:"name"`
	Shape    []int64 `json:"shape"`
	DataType string  `json:"datatype"`
	Data     any     `json:"data"`
}

// InferOutputParameter triton inference server infer parameters.
//type InferOutputParameter struct {
//	BinaryData     bool  `json:"binary_data,omitempty"`
//	Classification int64 `json:"classification,omitempty"`
//}

type Output struct {
	Data     []interface{} `json:"data"`
	Datatype string        `json:"datatype"`
	Name     string        `json:"name"`
	Shape    []int         `json:"shape"`
}

// HTTPOutput Model HTTP Request Output Struct.
type HTTPOutput struct {
	//Name       string               `json:"name"`
	//Parameters InferOutputParameter `json:"parameters,omitempty"`
	Outputs      []Output               `json:"outputs"`
}

// HTTPRequestBody Model HTTP Request Body.
type HTTPRequestBody struct {
	Inputs  []HTTPBatchInput `json:"inputs"`
	Outputs []HTTPOutput     `json:"outputs"`
}

func (m *SetFitModelService) getSetFitInputFeature(inferData string) (*InputFeature, *InputObjects) {
	// Replace BertDataSplitString Here, so the parts is 1, no need to use strings.Split and decrease a for-loop.
	if strings.Index(inferData, DataSplitString) > 0 {
		inferData = strings.ReplaceAll(inferData, DataSplitString, "")
	}
	// InputFeature
	// feature.TypeIDs  == segment_ids
	// feature.TokenIDs == input_ids
	// feature.Mask     == input_mask
	feature := &InputFeature{
		Text:   make([]string, m.MaxSeqLength),
	}
	inputObjects := &InputObjects{Input: inferData}
	return feature, inputObjects
}

// generateHTTPInputs get bert input feature for http request
// inferDataArr: model infer data slice
// inferInputs: triton inference server input tensor.
func (m *SetFitModelService) generateHTTPInputs(
	inferDataArr []string, inferInputs []*nvidia_inferenceserver.ModelInferRequest_InferInputTensor,
) ([]HTTPBatchInput, []*InputObjects) {
	// Feature
	batchModelInputObjs := make([]*InputObjects, len(inferDataArr))
	batchRequestInputs := make([]HTTPBatchInput, len(inferInputs))

	inferDataObjs := make([][][]string, len(inferDataArr))
	for i := range inferDataArr {
		feature, inputObject := m.getSetFitInputFeature(inferDataArr[i])
		batchModelInputObjs[i] = inputObject
		inferDataObjs[i] = [][]string{feature.Text}
	}
	inferDataObjs = utils.SliceTransposeFor3D(inferDataObjs)
	for i := range inferInputs {
		batchRequestInputs[i] = HTTPBatchInput{
			Name:     inferInputs[i].Name,
			Shape:    []int64{int64(len(inferDataArr)), int64(m.MaxSeqLength)},
			DataType: inferInputs[i].Datatype,
			Data:     inferDataObjs[i],
		}
	}
	return batchRequestInputs, batchModelInputObjs
}

// generateHTTPOutputs For HTTP Output.
func (m *SetFitModelService) generateHTTPOutputs(
	inferOutputs []*nvidia_inferenceserver.ModelInferRequest_InferRequestedOutputTensor,
) []HTTPOutput {
	requestOutputs := make([]HTTPOutput, len(inferOutputs))

	for i := range inferOutputs {
		fmt.Printf("%v\n", inferOutputs[i])
		//requestOutputs[i] = HTTPOutput{Outputs: inferOutputs[i].Outputs}
	}

	return requestOutputs
}

// generateHTTPRequest HTTP Request Data Generate.
func (m *SetFitModelService) generateHTTPRequest(
	inferDataArr []string,
	inferInputs []*nvidia_inferenceserver.ModelInferRequest_InferInputTensor,
	inferOutputs []*nvidia_inferenceserver.ModelInferRequest_InferRequestedOutputTensor,
) ([]byte, []*InputObjects, error) {
	// Generate batch request json body
	requestInputBody, modelInputObj := m.generateHTTPInputs(inferDataArr, inferInputs)
	jsonBody, jsonEncodeErr := m.TritonService.JsonMarshal(&HTTPRequestBody{
		Inputs:  requestInputBody,
		Outputs: m.generateHTTPOutputs(inferOutputs),
	})
	if jsonEncodeErr != nil {
		return nil, nil, jsonEncodeErr
	}
	return jsonBody, modelInputObj, nil
}

func (m *SetFitModelService) ModelInfer(
	inferData []string,
	modelName, modelVersion string,
	requestTimeout time.Duration,
	params ...interface{},
) ([]interface{}, error) {
	// Create request input/output tensors
	inferInputs := m.GenerateModelInferRequest()
	inferOutputs := m.GenerateModelInferOutputRequest(params...)

	httpRequestBody, httpInputData, err := m.generateHTTPRequest(inferData, inferInputs, inferOutputs)
	if err != nil {
		return nil, err
	}
	if httpRequestBody == nil {
		return nil, utils.ErrEmptyHTTPRequestBody
	}

	fmt.Println()
	// HTTP Infer
	return m.TritonService.ModelHTTPInfer(
		httpRequestBody, modelName, modelVersion, requestTimeout,
		m.InferCallback, m, httpInputData, params,
	)
}

func main() {
	model_name := "ensemble_model"
	model_version := "1"
	maxSeqLen := 128
        httpAddr := "localhost:8080"
	grpcAddr := "localhost:8001"
        httpClient := &fasthttp.Client{}

	defaultGRPCClient, grpcErr := grpc.NewClient(grpcAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if grpcErr != nil {
		panic(grpcErr)
	}

	modelService := models.ModelService{
		IsGRPC: false,
		MaxSeqLength:                    maxSeqLen,
		TritonService:                   nvidia_inferenceserver.NewTritonClientForAll(httpAddr, httpClient, defaultGRPCClient),
		InferCallback:                   testModerInferCallback,
		GenerateModelInferRequest:       testGenerateModelInferRequest,
		GenerateModelInferOutputRequest: testGenerateModelInferOutputRequest,
	}

	modelMetadataResponse, err := modelService.GetModelMeta(model_name, model_version, time.Second * 1)
	if err != nil {
		log.Fatalf(err.Error())
	}
	modelConfigResponse, err := modelService.GetModelConfig(model_name, model_version, time.Second * 1)
	if err != nil {
		log.Fatalf(err.Error())
	}

	modelReady, err := modelService.CheckModelReady(model_name, model_version, time.Second * 1)
	if err != nil {
		log.Fatalf(err.Error())
	}

	serverReady, err := modelService.CheckServerReady(time.Second * 1)
	if err != nil {
		log.Fatalf(err.Error())
	}

	fmt.Printf("modelMetadataResponse: %v\n", modelMetadataResponse)
	fmt.Printf("modelConfigResponse: %v\n", modelConfigResponse)
	fmt.Printf("modelReady: %v\n", modelReady)
	fmt.Printf("serverReady: %v\n", serverReady)

	m := NewSetFitModelService(modelService)

	// infer
	inferResultV1, err := m.ModelInfer([]string{"How to create policy?"}, model_name, model_version, 1*time.Second)
	if err != nil {
		log.Fatalf(err.Error())
	}
	fmt.Println(inferResultV1)
}
```

modelMetadataResponse: name:"ensemble_model" versions:"1" platform:"ensemble" inputs:{name:"TEXT" datatype:"BYTES" shape:-1} outputs:{name:"label" datatype:"BYTES" shape:-1} outputs:{name:"probabilities" datatype:"FP32" shape:-1 shape:2}
modelConfigResponse: config:{name:"ensemble_model" platform:"ensemble" version_policy:{latest:{num_versions:1}} input:{name:"TEXT" data_type:TYPE_STRING dims:-1} output:{name:"label" data_type:TYPE_STRING dims:-1} output:{name:"probabilities" data_type:TYPE_FP32 dims:-1 dims:2} ensemble_scheduling:{step:{model_name:"tokenize" model_version:-1 input_map:{key:"TEXT" value:"TEXT"} output_map:{key:"attention_mask" value:"attention_mask"} output_map:{key:"input_ids" value:"input_ids"} output_map:{key:"token_type_ids" value:"token_type_ids"}} step:{model_name:"python_setfit" model_version:-1 input_map:{key:"attention_mask" value:"attention_mask"} input_map:{key:"input_ids" value:"input_ids"} input_map:{key:"token_type_ids" value:"token_type_ids"} output_map:{key:"label" value:"tmp_label"} output_map:{key:"probabilities" value:"tmp_probabilities"}} step:{model_name:"label_mapping" model_version:-1 input_map:{key:"tmp_label" value:"tmp_label"} input_map:{key:"tmp_probabilities" value:"tmp_probabilities"} output_map:{key:"label" value:"label"} output_map:{key:"probabilities" value:"probabilities"}}} model_transaction_policy:{}}
modelReady: true
serverReady: true
name:"label"
name:"probabilities"

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x8cb181]

goroutine 1 [running]:
github.com/sunhailin-Leo/triton-service-go/v2/nvidia_inferenceserver.(*TritonClientService).httpErrorHandler(0xc0000ec3c0?, 0xc0000c4b00?, {0x0, 0x0})
	/home/ubuntu/go/pkg/mod/github.com/sunhailin-!leo/triton-service-go/v2@v2.0.2/nvidia_inferenceserver/triton_service_interface.go:236 +0x41
github.com/sunhailin-Leo/triton-service-go/v2/nvidia_inferenceserver.(*TritonClientService).ModelHTTPInfer(0xc0000ec3c0, {0xc000256600, 0x1f8, 0x200}, {0x9ffcf4, 0xe}, {0xac1fa0, 0x1}, 0x3b9aca00, 0xa3acd0, ...)
	/home/ubuntu/go/pkg/mod/github.com/sunhailin-!leo/triton-service-go/v2@v2.0.2/nvidia_inferenceserver/triton_service_interface.go:295 +0x273
main.(*SetFitModelService).ModelInfer(0xc0000eb580, {0xc0001aff18, 0x1, 0x1}, {0x9ffcf4, 0xe}, {0xac1fa0, 0x1}, 0x3b9aca00, {0x0, ...})
	/home/ubuntu/triton_inference_server/local_client/go/main.go:225 +0x292
main.main()
	/home/ubuntu/triton_inference_server/local_client/go/main.go:280 +0x705

**Comment by sunhailin-Leo** [comment]
@geraldstanje 
* You may compare the differences when you use triton-service-go and normal HTTP request to triton

**Comment by geraldstanje** [comment]
@sunhailin-Leo it seems i got it to work - let me update here.

**Comment by sunhailin-Leo** [comment]
> @sunhailin-Leo it seems i got it to work - let me update here.

Is your code working fine?
